pkg/operator/storageclasscontroller/storageclasscontroller.go

```diff
@@ -0,0 +1,228 @@
+	infralister "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25/mo"
+)
+
+const (
+	infraGlobalName       = "cluster"
+	cloudConfigNamespace  = "openshift-config"
+	categoryName          = "container-orchestrator"
+	tagName               = "openshift"
+	policyName            = "openshift-storage-policy"
+	vim25Prefix           = "urn:vim25:"
+	DatastoreInfoProperty = "info"
+	SummaryProperty       = "summary"
+)
+
+// associatedTypesRaw defines types with which created category can be associated with.
+var associatedTypesRaw = []string{"StoragePod", "Datastore", "ResourcePool", "VirtualMachine"}
+
+type StorageClassController struct {
+	name            string
+	targetNamespace string
+	manifest        []byte
+	kubeClient      kubernetes.Interface
+	operatorClient  v1helpers.OperatorClient
+	configMapLister corelister.ConfigMapLister
+	secretLister    corelister.SecretLister
+	infraLister     infralister.InfrastructureLister
+	recorder        events.Recorder
+}
+
+type vSphereConnection struct {
+	client     *govmomi.Client
+	config     *vsphere.VSphereConfig
+	tagManager *tags.Manager
+}
+
+func NewStorageClassController(
+	name,
+	targetNamespace string,
+	manifest []byte,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	operatorClient v1helpers.OperatorClient,
+	configInformer configinformers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	configMapInformer := kubeInformers.InformersFor(cloudConfigNamespace).Core().V1().ConfigMaps()
+	secretInformer := kubeInformers.InformersFor(targetNamespace).Core().V1().Secrets()
+	infraInformer := configInformer.Config().V1().Infrastructures()
+	rc := recorder.WithComponentSuffix("vmware-" + strings.ToLower(name))
+	c := &StorageClassController{
+		name:            name,
+		targetNamespace: targetNamespace,
+		manifest:        manifest,
+		kubeClient:      kubeClient,
+		operatorClient:  operatorClient,
+		configMapLister: configMapInformer.Lister(),
+		secretLister:    secretInformer.Lister(),
+		infraLister:     infraInformer.Lister(),
+		recorder:        rc,
+	}
+	return factory.New().WithInformers(
+		configMapInformer.Informer(),
+		secretInformer.Informer(),
+		infraInformer.Informer(),
+		operatorClient.Informer(),
+	).WithSync(
+		c.sync,
+	).ResyncEvery(
+		time.Minute, // TODO: figure out what's the idead resync time for this controller
+	).WithSyncDegradedOnError(
+		operatorClient,
```

Comment:
Do we really want to degrade the operator if we can't create the SC?

We could take a conservative approach for this operator and don't degrade. At least until we are more familiar how this operator behaves on the field.

Reply:
hmm, is there a reason you think  - creating storageclass would fail in a healthy cluster? With storage policy, it was bit more questionable because we needed more permissions etc. But with this code, I think degrading cluster is fine.

Reply:
Ack. I was more concerned about a hypothetical use-case where the cluster doesn't have any datastore. I'm far from being familiar with vSphere, so I don't even know if that's possible.

I'm fine with degrading the operator if such scenario is unlikely to exist.